### PR TITLE
4437 - shift subsequent project steps

### PR DESCRIPTION
**Pull request by elliottjf**
modal confirmation and logic to shift out subsequent step schedule dates when prior step updated as late

### CODE REVIEW

File: app/models/project_step.rb [hunk 1]
```diff
@@ -241,6 +242,34 @@ def finalize
     end
   end
 
+  # Returns number of days that the scheduled date has been moved out if finalized, or days late
+  # if just now marked completed.  Assumes that record has pending changes assigned, but not yet
+  # saved. Only returns 0 or a positive value.  Used to trigger potential automatic scheduled date
+  # shift of subsequent steps.
+  def pending_days_shifted
```
Comment (hooverlunch) [inline app/models/project_step.rb]: Great comment!

File: app/views/admin/project_steps/_project_step.html.slim [hunk 2]
```diff
@@ -9,6 +9,15 @@ div.panel.step style="border-color: #{step.border_color}" data-step-id="#{step.i
...
   = render "admin/project_steps/duplicate_step_modal", step: step if policy(step).duplicate?
 
+  - days_shifted ||= 0
+  - if days_shifted > 0
+    = render "admin/project_steps/shift_subsequent_confirm_modal", step: step,
+      days_shifted: days_shifted, subsequent_count: subsequent_count
+    javascript:
+      $(function () {
+        $("#step-shift-subsequent-confirm-modal-#{step.id}").modal("show");
+      });
```
Comment (hooverlunch) [inline app/views/admin/project_steps/_project_step.html.slim]: This JS could be in the partial, no?

File: app/views/admin/project_steps/_shift_subsequent_confirm_modal.html.slim [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+
+      div.modal-body
+        = t("project_step.confirm_shift_subsequent", count: subsequent_count, days: days_shifted)
+
+      div.modal-footer
+        button.btn.btn-default type="button" data-dismiss="modal"
+          = t(:no)
+
+        = link_to(t(:yes), shift_subsequent_admin_project_step_path(@step, num_days: days_shifted),
+          method: :patch, class: "btn btn-primary")
```
Comment (hooverlunch) [inline app/views/admin/project_steps/_shift_subsequent_confirm_modal.html.slim]: Newline pls.

File: app/models/project_step.rb [hunk 4]
```diff
@@ -241,6 +242,34 @@ def finalize
...
+    end
+    return [0,days_shifted].max
+  end
+
+  def subsequent_step_ids(previous_scheduled_date = nil)
+    date = previous_scheduled_date || scheduled_date
+    return [] unless date
+    # Todo: Confirm if this is the exact criteria desired.  It's unlikely that there would be
+    # prior uncompleted steps, but if there are perhaps we should just shift all uncompleted
+    # steps.
```
Comment (hooverlunch) [inline app/models/project_step.rb]: I'm sure this is fine for now.

File: app/views/admin/project_steps/_project_step.html.slim (hunk 2 above)
Comment (elliottjf) [inline app/views/admin/project_steps/_project_step.html.slim]: Sure.  Just to confirm, you're suggesting to move not literally just the JS block but the whole 'days shifted' block, lines 12-19?

Comment (hooverlunch) [inline app/views/admin/project_steps/_project_step.html.slim]: No, just the JS block. I also realized you can combine lines 12-13 (`if days_shifted && days_shifted > 0`).

**Comment by hooverlunch** [comment]
Excellent work. Thanks for this. I tweaked the wording of some things in the en.yml file. Pretty much good to go, just some very small tweaks.

**Pull request merged**